File: dock_spawn/decorators/resizable_container.dart
part of dock_spawn;

/**
 * Decorates a dock container with resizer handles around its base element
 * This enables the container to be resized from all directions
 */
class ResizableContainer implements IDockContainer {
  String containerType;
  IDockContainer delegate;
  Element topLevelElement;
  Dialog dialog;
  List<ResizeHandle> resizeHandles;
  
  Point2 dragOffset;
  Point2 previousMousePosition;
  
  ResizableContainer(this.dialog, this.delegate, this.topLevelElement) {
    containerType = delegate.containerType;
    topLevelElement.style.marginLeft = "${topLevelElement.offsetLeft}";
    topLevelElement.style.marginTop = "${topLevelElement.offsetTop}";
    _buildResizeHandles();
  }
  
  void setActiveChild(IDockContainer child) {
  }

  int get minimumAllowedChildNodes { return delegate.minimumAllowedChildNodes; }
  
  void _buildResizeHandles() {
    resizeHandles = new List<ResizeHandle>();
//    _buildResizeHandle(true, false, true, false); // Dont need the corner resizer near the close button
    _buildResizeHandle(false, true, true, false);
    _buildResizeHandle(true, false, false, true);
    _buildResizeHandle(false, true, false, true);

    _buildResizeHandle(true, false, false, false);
    _buildResizeHandle(false, true, false, false);
    _buildResizeHandle(false, false, true, false);
    _buildResizeHandle(false, false, false, true);

}
  
  void _buildResizeHandle(bool east, bool west, bool north, bool south) {
    var handle = new ResizeHandle();
    handle.east = east;
    handle.west = west;
    handle.north = north;
    handle.south = south;
    
    // Create an invisible div for the handle
    handle.element = new DivElement();
    topLevelElement.nodes.add(handle.element);
    
    // Build the class name for the handle
    String verticalClass = "";
    String horizontalClass = "";
    if (north) verticalClass = "n";
    if (south) verticalClass = "s";
    if (east) horizontalClass = "e";
    if (west) horizontalClass = "w";
    String cssClass = "resize-handle-$verticalClass$horizontalClass";
    if (verticalClass.length > 0 && horizontalClass.length > 0) {
      handle.corner = true;
    }

    handle.element.classes.add(handle.corner ? "resize-handle-corner" : "resize-handle");
    handle.element.classes.add(cssClass);
    resizeHandles.add(handle);
    
    // Create the mouse event handlers
    handle.mouseMoveHandler = (MouseEvent e) {
      onMouseMoved(handle, e);
    };
    handle.mouseDownHandler = (MouseEvent e) {
      onMouseDown(handle, e);
    };
    handle.mouseUpHandler = (MouseEvent e) {
      onMouseUp(handle, e);
    };

    handle.element.on.mouseDown.add(handle.mouseDownHandler);
  }

  void saveState(Map<String, Object> state) {
    delegate.saveState(state);
  }
  
  void loadState(Map<String, Object> state) {
    delegate.loadState(state);
  }
  
  int get width {
    return delegate.width;
  }

  int get height {
    return delegate.height;
  }

  String get name {
    return delegate.name;
  }
  void set name(String value) {
    delegate.name = value;
  }
  
  void resize(int _width, int _height) {
    delegate.resize(_width, _height);
    _adjustResizeHandles(_width, _height);
  }
  
  void _adjustResizeHandles(int _width, int _height) {
    resizeHandles.forEach((handle) {
      handle.adjustSize(topLevelElement, _width, _height);
    });
  }
  
  void performLayout(List<IDockContainer> children) {
    delegate.performLayout(children);
  }
  
  void destroy() {
    removeDecorator();
    delegate.destroy();
  }
  
  void removeDecorator() {
    
  }
    
  Element get containerElement {
    return delegate.containerElement;
  }

  bool readyToProcessNextResize = true;
  void onMouseMoved(ResizeHandle handle, MouseEvent e) {
    if (!readyToProcessNextResize) {
      return;
    }
    readyToProcessNextResize = false;
    
    window.requestLayoutFrame(() {
      Point2 currentMousePosition = new Point2(e.pageX, e.pageY);
      int dx = (currentMousePosition.x - previousMousePosition.x).toInt();
      int dy = (currentMousePosition.y - previousMousePosition.y).toInt();
      _performDrag(handle, dx, dy);
      previousMousePosition = currentMousePosition;
      readyToProcessNextResize = true;
    });
  }
  
  void onMouseDown(ResizeHandle handle, MouseEvent event) {
    previousMousePosition = new Point2(event.pageX, event.pageY);
    window.on.mouseMove.add(handle.mouseMoveHandler);
    window.on.mouseUp.add(handle.mouseUpHandler);

    document.body.classes.add("disable-selection");
  }
  
  void onMouseUp(ResizeHandle handle, MouseEvent event) {
    window.on.mouseMove.remove(handle.mouseMoveHandler);
    window.on.mouseUp.remove(handle.mouseUpHandler);

    document.body.classes.remove("disable-selection");
  }
  
  
  void _performDrag(ResizeHandle handle, int dx, int dy) {
    var bounds = new BoundingBox();
    bounds.left = getPixels(topLevelElement.style.marginLeft);
    bounds.top = getPixels(topLevelElement.style.marginTop);
    bounds.width = topLevelElement.clientWidth;
    bounds.height = topLevelElement.clientHeight;
    
    if (handle.east) _resizeEast(dx, bounds);
    if (handle.west) _resizeWest(dx, bounds);
    if (handle.north) _resizeNorth(dy, bounds);
    if (handle.south) _resizeSouth(dy, bounds);
  }
  
  void _resizeWest(int dx, BoundingBox bounds) {
    _resizeContainer(dx, 0, -dx, 0, bounds);
  }
  
  void _resizeEast(int dx, BoundingBox bounds) {
    _resizeContainer(0, 0, dx, 0, bounds);
  }

  void _resizeNorth(int dy, BoundingBox bounds) {
    _resizeContainer(0, dy, 0, -dy, bounds);
  }

  void _resizeSouth(int dy, BoundingBox bounds) {
    _resizeContainer(0, 0, 0, dy, bounds);
  }
  
  void _resizeContainer(int leftDelta, int topDelta, int widthDelta, int heightDelta, BoundingBox bounds) {
    bounds.left += leftDelta;
    bounds.top += topDelta;
    bounds.width += widthDelta;
    bounds.height += heightDelta;
    
    int minWidth = 50;  // TODO: Move to external configuration
    int minHeight = 50;  // TODO: Move to external configuration
    bounds.width = max(bounds.width, minWidth);
    bounds.height = max(bounds.height, minHeight);
    
    topLevelElement.style.marginLeft = "${bounds.left}px";
    topLevelElement.style.marginTop = "${bounds.top}px";

    resize(bounds.width, bounds.height);
  }
}

class ResizeHandle {
  final int handleSize = 6;   // TODO: Get this from DOM
  final int cornerSize = 12;  // TODO: Get this from DOM
  
  DivElement element;
  var mouseMoveHandler;
  var mouseDownHandler;
  var mouseUpHandler;
  bool east = false;
  bool west = false;
  bool north = false;
  bool south = false;
  bool corner = false;
  
  void adjustSize(Element container, int clientWidth, int clientHeight) {
    if (corner) {
      if (west) element.style.left = "0px";
      if (east) element.style.left = "${clientWidth - cornerSize}px";
      if (north) element.style.top = "0px";
      if (south) element.style.top = "${clientHeight - cornerSize}px";
    }
    else {
      if (west) {
        element.style.left = "0px";
        element.style.top = "${cornerSize}px";
      }
      if (east) {
        element.style.left = "${clientWidth - handleSize}px";
        element.style.top = "${cornerSize}px";
      }
      if (north) {
        element.style.left = "${cornerSize}px";
        element.style.top = "0px";
      }
      if (south) {
        element.style.left = "${cornerSize}px";
        element.style.top = "${clientHeight - handleSize}px";
      }
      
      if (west || east) {
        element.style.height = "${clientHeight - cornerSize * 2}px";
      } else {
        element.style.width = "${clientWidth - cornerSize * 2}px";
      }
    }
  }
}

File: js/lib/decorators/ResizableContainer.js
/**
 * Decorates a dock container with resizer handles around its base element
 * This enables the container to be resized from all directions
 */
dockspawn.ResizableContainer = function(dialog, delegate, topLevelElement)
{
    this.dialog = dialog;
    this.delegate = delegate;
    this.containerElement = delegate.containerElement;
    this.dockManager = delegate.dockManager;
    this.topLevelElement = topLevelElement;
    this.containerType = delegate.containerType;
    this.topLevelElement.style.marginLeft = this.topLevelElement.offsetLeft + "px";
    this.topLevelElement.style.marginTop = this.topLevelElement.offsetTop + "px";
    this.minimumAllowedChildNodes = delegate.minimumAllowedChildNodes;
    this._buildResizeHandles();
    this.readyToProcessNextResize = true;
};

dockspawn.ResizableContainer.prototype.setActiveChild = function(child)
{
};

dockspawn.ResizableContainer.prototype._buildResizeHandles = function()
{
    this.resizeHandles = [];
//    this._buildResizeHandle(true, false, true, false); // Dont need the corner resizer near the close button
    this._buildResizeHandle(false, true, true, false);
    this._buildResizeHandle(true, false, false, true);
    this._buildResizeHandle(false, true, false, true);

    this._buildResizeHandle(true, false, false, false);
    this._buildResizeHandle(false, true, false, false);
    this._buildResizeHandle(false, false, true, false);
    this._buildResizeHandle(false, false, false, true);
};

dockspawn.ResizableContainer.prototype._buildResizeHandle = function(east, west, north, south)
{
    var handle = new ResizeHandle();
    handle.east = east;
    handle.west = west;
    handle.north = north;
    handle.south = south;

    // Create an invisible div for the handle
    handle.element = document.createElement('div');
    this.topLevelElement.appendChild(handle.element);

    // Build the class name for the handle
    var verticalClass = "";
    var horizontalClass = "";
    if (north) verticalClass = "n";
    if (south) verticalClass = "s";
    if (east) horizontalClass = "e";
    if (west) horizontalClass = "w";
    var cssClass = "resize-handle-" + verticalClass + horizontalClass;
    if (verticalClass.length > 0 && horizontalClass.length > 0)
        handle.corner = true;

    handle.element.classList.add(handle.corner ? "resize-handle-corner" : "resize-handle");
    handle.element.classList.add(cssClass);
    this.resizeHandles.push(handle);

    var self = this;
    handle.mouseDownHandler = new dockspawn.EventHandler(handle.element, 'mousedown', function(e) { self.onMouseDown(handle, e); });
};

dockspawn.ResizableContainer.prototype.saveState = function(state)
{
    this.delegate.saveState(state);
};

dockspawn.ResizableContainer.prototype.loadState = function(state)
{
    this.delegate.loadState(state);
};

Object.defineProperty(dockspawn.ResizableContainer.prototype, "width", {
    get: function() { return this.delegate.width; }
});

Object.defineProperty(dockspawn.ResizableContainer.prototype, "height", {
    get: function() { return this.delegate.height; }
});

dockspawn.ResizableContainer.prototype.name = function(value)
{
    if (value)
        this.delegate.name = value;
    return this.delegate.name;
};

dockspawn.ResizableContainer.prototype.resize = function(width, height)
{
    this.delegate.resize(width, height);
    this._adjustResizeHandles(width, height);
};

dockspawn.ResizableContainer.prototype._adjustResizeHandles = function(width, height)
{
    var self = this;
    this.resizeHandles.forEach(function(handle) {
        handle.adjustSize(self.topLevelElement, width, height);
    });
};

dockspawn.ResizableContainer.prototype.performLayout = function(children)
{
    this.delegate.performLayout(children);
};

dockspawn.ResizableContainer.prototype.destroy = function()
{
    this.removeDecorator();
    this.delegate.destroy();
};

dockspawn.ResizableContainer.prototype.removeDecorator = function()
{
};

dockspawn.ResizableContainer.prototype.onMouseMoved = function(handle, e)
{
    if (!this.readyToProcessNextResize)
        return;
    this.readyToProcessNextResize = false;

//    window.requestLayoutFrame(() {
    this.dockManager.suspendLayout();
    var currentMousePosition = new Point(e.pageX, e.pageY);
    var dx = Math.floor(currentMousePosition.x - this.previousMousePosition.x);
    var dy = Math.floor(currentMousePosition.y - this.previousMousePosition.y);
    this._performDrag(handle, dx, dy);
    this.previousMousePosition = currentMousePosition;
    this.readyToProcessNextResize = true;
    this.dockManager.resumeLayout();
//    });
};

dockspawn.ResizableContainer.prototype.onMouseDown = function(handle, event)
{
    this.previousMousePosition = new Point(event.pageX, event.pageY);
    if (handle.mouseMoveHandler)
    {
        handle.mouseMoveHandler.cancel();
        delete handle.mouseMoveHandler
    }
    if (handle.mouseUpHandler)
    {
        handle.mouseUpHandler.cancel();
        delete handle.mouseUpHandler
    }

    // Create the mouse event handlers
    var self = this;
    handle.mouseMoveHandler = new dockspawn.EventHandler(window, 'mousemove', function(e) { self.onMouseMoved(handle, e); });
    handle.mouseUpHandler = new dockspawn.EventHandler(window, 'mouseup', function(e) { self.onMouseUp(handle, e); });

    document.body.classList.add("disable-selection");
};

dockspawn.ResizableContainer.prototype.onMouseUp = function(handle, event)
{
    handle.mouseMoveHandler.cancel();
    handle.mouseUpHandler.cancel();
    delete handle.mouseMoveHandler;
    delete handle.mouseUpHandler;

    document.body.classList.remove("disable-selection");
};

dockspawn.ResizableContainer.prototype._performDrag = function(handle, dx, dy)
{
    var bounds = {};
    bounds.left = getPixels(this.topLevelElement.style.marginLeft);
    bounds.top = getPixels(this.topLevelElement.style.marginTop);
    bounds.width = this.topLevelElement.clientWidth;
    bounds.height = this.topLevelElement.clientHeight;

    if (handle.east) this._resizeEast(dx, bounds);
    if (handle.west) this._resizeWest(dx, bounds);
    if (handle.north) this._resizeNorth(dy, bounds);
    if (handle.south) this._resizeSouth(dy, bounds);
};

dockspawn.ResizableContainer.prototype._resizeWest = function(dx, bounds)
{
    this._resizeContainer(dx, 0, -dx, 0, bounds);
};

dockspawn.ResizableContainer.prototype._resizeEast = function(dx, bounds)
{
    this._resizeContainer(0, 0, dx, 0, bounds);
};

dockspawn.ResizableContainer.prototype._resizeNorth = function(dy, bounds)
{
    this._resizeContainer(0, dy, 0, -dy, bounds);
};

dockspawn.ResizableContainer.prototype._resizeSouth = function(dy, bounds)
{
    this._resizeContainer(0, 0, 0, dy, bounds);
};

dockspawn.ResizableContainer.prototype._resizeContainer = function(leftDelta, topDelta, widthDelta, heightDelta, bounds)
{
    bounds.left += leftDelta;
    bounds.top += topDelta;
    bounds.width += widthDelta;
    bounds.height += heightDelta;

    var minWidth = 50;  // TODO: Move to external configuration
    var minHeight = 50;  // TODO: Move to external configuration
    bounds.width = Math.max(bounds.width, minWidth);
    bounds.height = Math.max(bounds.height, minHeight);

    this.topLevelElement.style.marginLeft = bounds.left + "px";
    this.topLevelElement.style.marginTop = bounds.top + "px";

    this.resize(bounds.width, bounds.height);
};


function ResizeHandle()
{
    this.element = undefined;
    this.handleSize = 6;   // TODO: Get this from DOM
    this.cornerSize = 12;  // TODO: Get this from DOM
    this.east = false;
    this.west = false;
    this.north = false;
    this.south = false;
    this.corner = false;
}

ResizeHandle.prototype.adjustSize = function(container, clientWidth, clientHeight)
{
    if (this.corner)
    {
        if (this.west) this.element.style.left = "0px";
        if (this.east) this.element.style.left = (clientWidth - this.cornerSize) + "px";
        if (this.north) this.element.style.top = "0px";
        if (this.south) this.element.style.top = (clientHeight - this.cornerSize) + "px";
    }
    else
    {
        if (this.west)
        {
            this.element.style.left = "0px";
            this.element.style.top = this.cornerSize + "px";
        }
        if (this.east) {
            this.element.style.left = (clientWidth - this.handleSize) + "px";
            this.element.style.top = this.cornerSize + "px";
        }
        if (this.north) {
            this.element.style.left = this.cornerSize + "px";
            this.element.style.top = "0px";
        }
        if (this.south) {
            this.element.style.left = this.cornerSize + "px";
            this.element.style.top = (clientHeight - this.handleSize) + "px";
        }

        if (this.west || this.east) {
            this.element.style.height = (clientHeight - this.cornerSize * 2) + "px";
        } else {
            this.element.style.width = (clientWidth - this.cornerSize * 2) + "px";
        }
    }
};
